extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-plugin-service-impl.ts

```diff
@@ -0,0 +1,739 @@
+    }
+  }
+
+  /**
+   * Fetches plugin meta.yaml
+   * @param plugin pluin in format publisher/name/version
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async getPluginMetaYaml(plugin: string): Promise<any> {
+    const registryURL = this.trimTrailingSlash(this.env.getPluginRegistryInternalURL());
+    const yamlURL = `${registryURL}/plugins/${plugin}/meta.yaml`;
+    const metaYaml = await this.httpService.get(yamlURL);
+
+    if (!metaYaml) {
+      const msg = `Unable to get ${yamlURL}`;
+      // console.error(msg);
+      throw new Error(msg);
+    }
+
+    return jsYaml.safeLoad(metaYaml);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async installPluginToEditorContainer(yaml: any): Promise<void> {
+    // get list of extensions
+    if (yaml.spec && yaml.spec.extensions) {
+      const downloadedExtensions: string[] = [];
+
+      const extensions = yaml.spec.extensions;
+      for (const extension of extensions) {
+        if (await this.isExtensionAlreadyDownloaded(extension)) {
+          continue;
+        }
+
+        const extensionFile = await this.downloadExtensionToPluginsDirectory(extension);
+        if (extensionFile) {
+          downloadedExtensions.push(extensionFile);
+        } else {
+          // rollback the installation, remove all downloaded files
+          for (const file of downloadedExtensions) {
+            await fs.remove(file);
+          }
+
+          throw new Error(`Cannot download ${extension}`);
+        }
+      }
+    }
+  }
+
+  async downloadExtensionToPluginsDirectory(url: string, sidecar?: string): Promise<string | undefined> {
+    let target;
+
+    if (sidecar) {
+      const directoryName = `${SIDECAR_PLUGINS_DIR}/${sidecar}`;
+      await fs.ensureDir(directoryName);
+      target = `${SIDECAR_PLUGINS_DIR}/${sidecar}/${path.basename(url)}`;
+    } else {
+      target = `${PLUGINS_DIR}/${path.basename(url)}`;
+    }
+
+    try {
+      const response = await this.httpService.get(url, 'arraybuffer');
+      if (response) {
+        await fs.writeFile(target, response, { encoding: 'binary' });
+        return target;
+      }
+    } catch (error) {
+      console.error(error.message);
+    }
+
+    return undefined;
+  }
+
+  async isExtensionAlreadyDownloaded(url: string, sidecar?: string): Promise<boolean> {
+    const target = sidecar
+      ? `${SIDECAR_PLUGINS_DIR}/${sidecar}/${path.basename(url)}`
+      : `${PLUGINS_DIR}/${path.basename(url)}`;
+    return fs.pathExists(target);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async installPluginToDeveloperContainer(yaml: any): Promise<void> {
+    const devfile: Devfile = await this.devfileService.get();
+
+    if (!yaml.spec.extensions) {
+      console.log('Plugin does not provide any extension.');
+    }
+
+    await this.injectExtensionsInDevContainer(yaml, devfile);
+
+    await this.devfileService.updateDevfile(devfile);
```

Comment:
we should patch the current object instead of updating the full devfile

example: https://github.com/eclipse-che/che-theia/blob/main/extensions/eclipse-che-theia-remote-impl-k8s/src/node/k8s-workspace-service-impl.ts#L93-L120

Reply:
Done